include/NovelRT/Maths/QuadTree.h

```diff
@@ -0,0 +1,135 @@
+#ifndef NOVELRT_MATHS_QUADTREE_H
+#define NOVELRT_MATHS_QUADTREE_H
+
+#ifndef NOVELRT_H
+#error Please do not include this directly. Use the centralised header (NovelRT.h) instead!
+#endif
+
+namespace NovelRT::Maths {
+  template <typename TQuadTreePoint>
+  class QuadTree {
+  private:
+    static const int32_t POINT_CAPACITY = 4;
+
+    static const int32_t TOP_LEFT = 0;
+    static const int32_t TOP_RIGHT = 1;
+    static const int32_t BOTTOM_LEFT = 2;
+    static const int32_t BOTTOM_RIGHT = 3;
+
+    typedef std::array<std::shared_ptr<TQuadTreePoint>, POINT_CAPACITY> TPoints;
+    typedef std::array<std::unique_ptr<QuadTree>, 4> TChildren;
+
+    GeoBounds _bounds;
+    TPoints _points;
+    TChildren _children;
+    uint32_t _pointCount;
+
+  public:
+    explicit QuadTree(const GeoBounds& bounds) noexcept :
+      _bounds(bounds),
+      _points({}),
+      _children({}),
+      _pointCount(0) {
+      static_assert(std::is_base_of<QuadTreePoint, TQuadTreePoint>::value, "Type argument for TQuadTreePoint must inherit NovelRT::Maths::QuadTreePoint!");
+    }
+
+    const GeoBounds& getBounds() const noexcept {
+      return _bounds;
+    }
+
+    std::shared_ptr<TQuadTreePoint> getPoint(uint32_t index) const noexcept {
+      return (index < _pointCount) ? _points[index] : nullptr;
+    }
+
+    uint32_t getPointCount() const noexcept {
+      return _pointCount;
+    }
+
+    QuadTree* getTopLeft() const noexcept {
+      return _children[TOP_LEFT].get();
+    }
+
+    QuadTree* getTopRight() const noexcept {
+      return _children[TOP_RIGHT].get();
+    }
+
+    QuadTree* getBottomLeft() const noexcept {
+      return _children[BOTTOM_LEFT].get();
+    }
+
+    QuadTree* getBottomRight() const noexcept {
+      return _children[BOTTOM_RIGHT].get();
+    }
+
+    bool tryInsert(std::shared_ptr<TQuadTreePoint> point) noexcept {
+      if (point == nullptr || !getBounds().pointIsWithinBounds(point->getPosition())) return false;
+
+      if (_children[TOP_LEFT] == nullptr) {
```

Comment:
this should be using `.at()` or `.find()` most probably?

Reply:
this is `std::array` so it is fixed size and won't risk accidentally growing. We are also using a known constant index, so we don't want `find()`.